### PR TITLE
Split os-release output according to backend type

### PR DESCRIPTION
**Pull request by mloviska**
VMWare and HyperV backends use CR LF instead of LF as it is in other
backends. To properly parse os-release file from SUT we need to choose
separator in runtime according to used backend.

- [test fails in docker_image and podman_image](https://progress.opensuse.org/issues/80968)
- VRs:
  * [Build20.160-jeos-container-engines_and_tools@uefi-virtio-vga](http://kepler.suse.cz/tests/2802#)
  * [Build20.160-jeos-container-engines_and_tools@svirt-vmware65](http://kepler.suse.cz/tests/2803#)
  * [Build20.160-jeos-container-engines_and_tools@svirt-hyperv-uefi](http://kepler.suse.cz/tests/2804#)

### CODE REVIEW

**Review by pdostal — COMMENTED** [review]
This is perfect catch 👏 I will do additional verification runs and then approve.

**Comment by mloviska** [comment]
> This is perfect catch clap I will do additional verification runs and then approve.

Thank you Pavliku, additional VRs are very much appreciated! If possible, could you please share with us a pcloud run ?

**Comment by Vogtinator** [comment]
If I get this right, the serial console on various platforms mangles newlines? That's really not good, as it means that `script_output` and `serial_terminal` and the likes are just not reliable and IMO not usable that way.

In `autoinst-log.txt` it's visible that other calls are equally affected.

Is there a way to fix this is the backend code instead?

**Comment by Vogtinator** [comment]
There's actually code for this in `wait_serial` already: https://github.com/os-autoinst/os-autoinst/blob/963434dcb2396afc7f395d0153cd57076a05fe0b/testapi.pm#L929

**Comment by mloviska** [comment]
> If I get this right, the serial console on various platforms mangles newlines? That's really not good, as it means that `script_output` and `serial_terminal` and the likes are just not reliable and IMO not usable that way.
> 
I guess the CR LF is somehow expected from `hyperv` as we use a named pipe TCP proxy to fetch the serial line output.
AFAICT `serial_terminal` is not usable anyway in `vmware`, `hyperv` and `xen` environments at the moment.

The initial first step should be enabling [serial_terminal for xen](https://progress.opensuse.org/issues/55985).

> In `autoinst-log.txt` it's visible that other calls are equally affected.
> 
> Is there a way to fix this is the backend code instead?

I haven't checked the backend code properly yet. As of now, we need to think of this issue whenever using `script_output` for vmware and hyperv jobs.
I would leave it as hotfix until we find a way how to fix it in the backend.

**Comment by mloviska** [comment]
> There's actually code for this in `wait_serial` already: https://github.com/os-autoinst/os-autoinst/blob/963434dcb2396afc7f395d0153cd57076a05fe0b/testapi.pm#L929

Yup, that was the other implementation what I thought of, on the other hand this way it required fewer changes. Basically, it is the same thing just more descriptive.

**Comment by Vogtinator** [comment]
> > There's actually code for this in `wait_serial` already: https://github.com/os-autoinst/os-autoinst/blob/963434dcb2396afc7f395d0153cd57076a05fe0b/testapi.pm#L929
> 
> Yup, that was the other implementation what I thought of, on the other hand this way it required fewer changes. Basically, it is the same thing just more descriptive.

What I'm saying is that `script_output` internally calls `wait_serial`, so this case should actually be handled already.

**Review by pdostal — APPROVED** [review]
More verification runs:
 * [Tumbleweed](http://pdostal-server.suse.cz/tests/10671)
 * [SLE12-SP5PC](http://pdostal-server.suse.cz/tests/10669)
 * [SLE15-SP2PC](http://pdostal-server.suse.cz/tests/10670)

File: lib/version_utils.pm [hunk 1]
```diff
@@ -578,9 +578,20 @@ sub get_os_release {
...
     $go_to_target    //= '';
     $os_release_file //= '/etc/os-release';
-    my %os_release = script_output("$go_to_target cat $os_release_file") =~ /^([^#]\S+)="?([^"\r\n]+)"?$/gm;
-    %os_release = map { uc($_) => $os_release{$_} } keys %os_release;
-    my ($os_version, $os_service_pack) = split(/\.|-sp/i, $os_release{VERSION});
+
+    my %os_release;
+    my $sep = (is_vmware || is_hyperv) ? "\r\n" : "\n";
+
+    foreach my $line (split(/$sep/, script_output("$go_to_target cat $os_release_file"))) {
```
Comment (Vogtinator) [inline lib/version_utils.pm:585]: The regex itself already handles the splitting, so just this change would be enough:

```
my %os_release = script_output("$go_to_target cat $os_release_file") =~ /^([^#]\S+)="?([^"\r\n]+)"?\r?$/gm;
```

**Review by Vogtinator — CHANGES_REQUESTED** [review]
IMO it should be investigated why the code in `wait_serial` does not work here first.

**Comment by mloviska** [comment]
> IMO it should be investigated why the code in `wait_serial` does not work here first.

Ok, I will have a look.

**Comment by mloviska** [comment]
> IMO it should be investigated why the code in `wait_serial` does not work here first.

If I haven't missed anything, the only `wait_serial` call in case of `vmware` and `hyperv` is executed on [distribution.pm#L278](https://github.com/os-autoinst/os-autoinst/blob/master/distribution.pm#L278) and it checks the script's RC.
Maybe we can trim the `CRLF` as it is done with white spaces at the end of [script_output](https://github.com/os-autoinst/os-autoinst/blob/master/distribution.pm#L294).

**Comment by Vogtinator** [comment]
> > IMO it should be investigated why the code in `wait_serial` does not work here first.
> 
> If I haven't missed anything, the only `wait_serial` call in case of `vmware` and `hyperv` is executed on [distribution.pm#L278](https://github.com/os-autoinst/os-autoinst/blob/master/distribution.pm#L278) and it checks the script's RC.

The return value of the call is used as the return value of `script_output`, note also the `record_output => 1`.

**Comment by mloviska** [comment]
> > > IMO it should be investigated why the code in `wait_serial` does not work here first.
> > 
> > 
> > If I haven't missed anything, the only `wait_serial` call in case of `vmware` and `hyperv` is executed on [distribution.pm#L278](https://github.com/os-autoinst/os-autoinst/blob/master/distribution.pm#L278) and it checks the script's RC.
> 
> The return value of the call is used as the return value of `script_output`, note also the `record_output => 1`.

That should affect `serial_terminal` suitable environments only.

```
[2020-12-11T08:42:06.369 CET] [debug] tests/console/consoletest_setup.pm:35 called opensusebasetest::select_serial_terminal -> lib/opensusebasetest.pm:1224 called testapi::select_console -> lib/susedistribution.pm:716 called serial_terminal::login -> lib/serial_terminal.pm:109 called testapi::wait_serial
[2020-12-11T08:42:06.369 CET] [debug] <<< testapi::wait_serial(expect_not_found=0, regexp=qr/login:\s*$/ui, quiet=1, no_regex=0, timeout=5, record_output=undef, buffer_size=undef)
[2020-12-11T08:42:06.369 CET] [debug] <<< consoles::serial_screen::read_until(json_cmd_token="qAxIrGKi", buffer_size=undef, record_output=undef, timeout=5, cmd="backend_wait_serial", no_regex=0, quiet=1, expect_not_found=0, pattern="(?^ui:login:\\s*\$)", regexp="(?^ui:login:\\s*\$)")
[33m[2020-12-11T08:42:06.370 CET] [info] ::: consoles::serial_screen::read_until: Matched output from SUT in 2 loops & 0.000255911960266531 seconds: login: 
[0m[32m[2020-12-11T08:42:06.370 CET] [debug] >>> testapi::wait_serial: (?^ui:login:\s*$): ok
```

**Comment by pdostal** [comment]
* [SLE based image on Ubuntu](http://pdostal-server.suse.cz/tests/10677)
 * [SLE based image on Centos](http://pdostal-server.suse.cz/tests/10675)

**Comment by Vogtinator** [comment]
Looks like the `static.yml` file has to be updated to account for https://github.com/os-autoinst/os-autoinst/commit/a07a1cb3f4a3b1c62d2492bcab13215615191dd5. @okurz

**Comment by mloviska** [comment]
> Looks like the `static.yml` file has to be updated to account for [os-autoinst/os-autoinst@a07a1cb](https://github.com/os-autoinst/os-autoinst/commit/a07a1cb3f4a3b1c62d2492bcab13215615191dd5). @okurz

https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11602

**Comment by Vogtinator** [comment]
Heh, https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11604#issuecomment-743154972 was actually meant for a different PR (pasted into wrong tab...), but apparently it was still relevant :D

> That should affect serial_terminal suitable environments only.

I'm not following there, what I see is:

```
[2020-12-09T11:06:09.406 CET] [debug] tests/containers/podman.pm:42 called version_utils::get_os_release -> lib/version_utils.pm:581 called testapi::script_output
[2020-12-09T11:06:09.407 CET] [debug] <<< testapi::wait_serial(record_output=1, regexp="SCRIPT_FINISHEDn1~0x-\\d+-", timeout=90, buffer_size=undef, no_regex=0, quiet=undef, expect_not_found=0)
NAME="SLES"

VERSION="15-SP3 Beta2"
```

File: lib/version_utils.pm (hunk 1 above)
Comment (b10n1k) [inline lib/version_utils.pm:585]: seems to me that the only change that you need is 
```
my $sep = (is_vmware || is_hyperv) ? "\r\n" : "\n";                                                                                                                                                                                      
    my %os_release = script_output("$go_to_target cat $os_release_file") =~ /^([^#]\S+)="?([^"$sep]+)"?$/gm;
```
if i am not wrong. Why do you need a loop? in any case, in the way that is implemented now, i would suggest to move the `script_output` outside the loop and keep the `dial` . wdyt?

Comment (pdostal) [reply lib/version_utils.pm:585]: I find Martin's solution more readable. The previous solution is looping over the hash as well.

**Comment by mloviska** [comment]
> There's actually code for this in `wait_serial` already: https://github.com/os-autoinst/os-autoinst/blob/963434dcb2396afc7f395d0153cd57076a05fe0b/testapi.pm#L929

I have hardcoded newline character in split `foreach my $line (split(/\n/, script_output("$go_to_target cat $os_release_file")))` and 
modified `$ret->{string} =~ s,\r+,,g;`in testapi.pm.

[Build20.160-jeos-container-engines_and_tools@svirt-hyperv-uefi](http://kepler.suse.cz/tests/2842#downloads)

```
[0mTESToutput $VAR1 = 'n1~0x
NAME="SLES"
VERSION="15-SP3 Beta2"
VERSION_ID="15.3"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp3"
DOCUMENT_URL="https://documentation.suse.com/"
SCRIPT_FINISHEDn1~0x-0-
';
TESTout $VAR1 = '
NAME="SLES"
VERSION="15-SP3 Beta2"
VERSION_ID="15.3"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp3"
DOCUMENT_URL="https://documentation.suse.com/"
';
[37m[2020-12-11T17:08:20.563 CET] [debug] Parsing NAME="SLES" from /etc/os-release
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: key=NAME
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: value=SLES
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] Parsing VERSION="15-SP3 Beta2" from /etc/os-release
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: key=VERSION
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: value=15-SP3 Beta2
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] Parsing VERSION_ID="15.3" from /etc/os-release
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: key=VERSION_ID
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: value=15.3
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] Parsing PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)" from /etc/os-release
[0m[37m[2020-12-11T17:08:20.563 CET] [debug] get_os_release: key=PRETTY_NAME
```

instead of 

[Build20.160-jeos-container-engines_and_tools@svirt-hyperv-uefi](http://kepler.suse.cz/tests/2845/file/autoinst-log.txt)

```
[0mTESToutput $VAR1 = 'n1~0x
NAME="SLES"
VERSION="15-SP3 Beta2"
VERSION_ID="15.3"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp3"
DOCUMENT_URL="https://documentation.suse.com/"
SCRIPT_FINISHEDn1~0x-0-
';
TESTout $VAR1 = '
NAME="SLES"
VERSION="15-SP3 Beta2"
VERSION_ID="15.3"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp3"
DOCUMENT_URL="https://documentation.suse.com/"
';
[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing NAME="SLES"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing VERSION="15-SP3 Beta2"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing VERSION_ID="15.3"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing PRETTY_NAME="SUSE Linux Enterprise Server 15 SP3 (Beta2)"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing ID="sles"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing ID_LIKE="suse"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing ANSI_COLOR="0;32"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing CPE_NAME="cpe:/o:suse:sles:15:sp3"
 from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] Parsing DOCUMENT_URL="https://documentation.suse.com/" from /etc/os-release
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] get_os_release: key=DOCUMENT_URL
[0m[37m[2020-12-11T18:03:19.936 CET] [debug] get_os_release: value=https://documentation.suse.com/
```

**Comment by mloviska** [comment]
https://github.com/os-autoinst/os-autoinst/pull/1600

**Comment by cfconrad** [comment]
> I will do more tests later. It really looks like a problem in `wait_serial`, thanks Fabian for bringing it to table!
I had similar things with a "unfortunately" not accepted PR#1165.
Take a look to this line. We have backends which produce neighter `\n` nor `\r\n` but `\025` https://github.com/os-autoinst/os-autoinst/pull/1165/files#diff-5ee7172d501e1ba3014d720a7f4eeea649f679cc81f60c7373fec334116e85a9R278

So in our case, we could try increase the regex in wait_serial with `\025` like, maybe this helps here too.
```
    $ret->{string} =~ s/(\r\n|\025)/\n/g;
```

**Comment by cfconrad** [comment]
Just read now https://github.com/os-autoinst/os-autoinst/pull/1600, is this PR now obsolete?

**Comment by Vogtinator** [comment]
> Just read now [os-autoinst/os-autoinst#1600](https://github.com/os-autoinst/os-autoinst/pull/1600), is this PR now obsolete?

The issue which this PR was originally for should no longer be there, yes.

**Pull request closed without merging**